### DISCUSSION TITLE
Is there an easy way to make just one specific window floating and centered in the current monitor/workspace?

### DISCUSSION THREAD

**Discussion by ghost** [nikitabobko/AeroSpace#633] [questions-and-answers]
I tried using the following bash script (called with a keyboard shortcut that is then executed by Raycast):

```
#!/bin/bash

# Required parameters:
# @raycast.schemaVersion 1
# @raycast.title yekuake-omni
# @raycast.mode silent

# Optional parameters:
# @raycast.icon 🤖

nu -c "emacsclient -s /Users/oscarvarto/.local/run/emacs/server -n -e '(yequake-toggle "\""yequake-omni & scratch"\"")'" &&
  aerospace move left &&
  aerospace layout floating
```

It kind-of works, because it creates the emacs buffer and puts it horizontally centered for a specific starting state of my Windows (a terminal and an always present Emacs frame). _But_ Aerospace creates the window, modifies the tiling to make space for this newly created window, then moves it and changes the window to float.

The end result with all the window moving is very uncomfortable/distracting. Is there a way to suppress all the window animation/movements (sort of showing the end result window at a centered position only)? Maybe this is not possible, just wondering...

By the way, I am using "Reduce motion" ON in my `Accessibility > Display` setttings.

**Comment by ghost** [comment]
Related discussion: https://github.com/armindarvish/consult-omni/discussions/19#discussioncomment-10168740

**Comment by gldtn** [comment]
Looking for the same thing, in yabai, you can do something like: `yabai -m window --toggle float --grid 4:4:1:1:2:2` which centers the floating window to the screen. Is this possible with aerospace? I did not see anything in the doc related to this!

**Comment by maxpetretta** [comment]
also looking for something like this!

**Comment by vieitesss** [comment]
I've used bspwm extensively. It has a layout called "pseudo-tiling," which allows you to have a window in tiling mode with the ability to resize it individually. This feature lets you keep a single window centered at your desired size, and opening other windows will not overlap the first one.

I think this could be the solution to this issue.

**Comment by lambdalisue** [comment]
It's not **easy** and quite **fragile** so this is just a workaround, but I managed to achieve it using AppleScript. Just FYI.

I wish if this feature is natively supported by aerospace itself.

https://gist.github.com/lambdalisue/5f0ce431b0f41177633c0572fe47ccd7

https://github.com/user-attachments/assets/8ef1603c-7341-427d-97ac-e0fc56078904

**Comment by franckrasolo** [comment]
I took @lambdalisue's [gist](https://gist.github.com/lambdalisue/5f0ce431b0f41177633c0572fe47ccd7) ❤️ and [implemented](https://github.com/franckrasolo/dotfiles.nix/blob/d87978c18e42313f0c43026f5488646126f48e89/darwin/aerospace/aerospace.toml#L104-L120) the following features using AppleScript:

- toggling of a floating centered window: [script](https://github.com/franckrasolo/dotfiles.nix/blob/d87978c18e42313f0c43026f5488646126f48e89/darwin/aerospace/toggle-floating-centered.sh)
- resizing of a floating centered window: [script](https://github.com/franckrasolo/dotfiles.nix/blob/d87978c18e42313f0c43026f5488646126f48e89/darwin/aerospace/resize-floating-centered.sh)
- moving of a floating window within the screen boundaries: [script](https://github.com/franckrasolo/dotfiles.nix/blob/d87978c18e42313f0c43026f5488646126f48e89/darwin/aerospace/move-floating.sh)

**Reply by patroza** [reply]
this is cool, however when executed from Aerospace it just doesn't do anything. seems to have to do with accessibility permissions but no matter which apps I add there, it doesn't seem to help. only when I execute it manually from terminal it works :S

**Comment by PhenomenaPh** [comment]
Any updates on this? Seems like a basic feature needed to be included

**Comment by wojciech-kulik** [comment]
+1 there should be a command to set a fixed position of a floating window

**Comment by gldtn** [comment]
I agree it would be a nice implementation to have Aerospace supported natively. Meanwhile since I don't know swift, I attempted to use Chatgpt to write a swift script to take care of this and it's feels much faster than a shell script. So full credit for AI for getting this working properly 😬  

center_floats.swift:
```swift
import AppKit

// Function to parse arguments
func parseArguments() -> (width: CGFloat, height: CGFloat) {
    var width: CGFloat = 1600
    var height: CGFloat = 900

    let args = CommandLine.arguments
    for (index, arg) in args.enumerated() {
        switch arg {
        case "-w":
            if index + 1 < args.count, let w = Double(args[index + 1]) {
                width = CGFloat(w)
            }
        case "-h":
            if index + 1 < args.count, let h = Double(args[index + 1]) {
                height = CGFloat(h)
            }
        default:
            break
        }
    }
    return (width, height)
}

// Parse command-line arguments
let (desiredWidth, desiredHeight) = parseArguments()

// Function to run a shell command
@discardableResult
func runShellCommand(_ command: String) -> Int32 {
    let process = Process()
    process.executableURL = URL(fileURLWithPath: "/bin/zsh") // Change to /bin/bash if using bash
    process.arguments = ["-c", command]
    process.launch()
    process.waitUntilExit()
    return process.terminationStatus
}

// Function to get the frontmost window
func getFrontmostWindow() -> AXUIElement? {
    guard let frontAppPID = NSWorkspace.shared.frontmostApplication?.processIdentifier else {
        print("Failed to get the frontmost application.")
        return nil
    }
    let appElement = AXUIElementCreateApplication(frontAppPID)
    var focusedWindow: CFTypeRef?
    let result = AXUIElementCopyAttributeValue(appElement, kAXFocusedWindowAttribute as CFString, &focusedWindow)
    if result == .success, let window = focusedWindow {
        return (window as! AXUIElement)
    } else {
        print("Failed to get the focused window.")
        return nil
    }
}

// Function to get the monitor where the window is currently located
func getMonitorForWindow(window: AXUIElement) -> NSScreen? {
    var windowFrame = CGRect.zero
    var value: CFTypeRef?

    // Get window position
    if AXUIElementCopyAttributeValue(window, kAXPositionAttribute as CFString, &value) == .success,
       AXValueGetValue(value as! AXValue, .cgPoint, &windowFrame.origin) {
        // Get window size
        if AXUIElementCopyAttributeValue(window, kAXSizeAttribute as CFString, &value) == .success,
           AXValueGetValue(value as! AXValue, .cgSize, &windowFrame.size) {
            // Find the screen containing this window
            return NSScreen.screens.first(where: { $0.frame.intersects(windowFrame) })
        }
    }
    return nil
}

// Function to set window position and size
func setWindowPositionAndSize(window: AXUIElement, x: CGFloat, y: CGFloat, width: CGFloat, height: CGFloat) {
    var position = CGPoint(x: x, y: y) // Create mutable CGPoint
    var size = CGSize(width: width, height: height) // Create mutable CGSize
    let positionValue = AXValueCreate(.cgPoint, &position) // Pass mutable variable
    let sizeValue = AXValueCreate(.cgSize, &size) // Pass mutable variable
    if let positionValue = positionValue, let sizeValue = sizeValue {
        AXUIElementSetAttributeValue(window, kAXPositionAttribute as CFString, positionValue)
        AXUIElementSetAttributeValue(window, kAXSizeAttribute as CFString, sizeValue)
    }
}

// Main execution
// Run Aerospace commands
let flattenStatus = runShellCommand(let flattenStatus = runShellCommand("aerospace layout floating || aerospace layout tiling")
if flattenStatus != 0 {
    print("Failed to execute Aerospace commands.")
    exit(1)
}

if let window = getFrontmostWindow() {
    if let screen = getMonitorForWindow(window: window) ?? NSScreen.main {
        let screenFrame = screen.visibleFrame
        let newX = screenFrame.origin.x + (screenFrame.width - desiredWidth) / 2
        let newY = screenFrame.origin.y + (screenFrame.height - desiredHeight) / 2
        setWindowPositionAndSize(window: window, x: newX, y: newY, width: desiredWidth, height: desiredHeight)
    } else {
        print("Failed to get the monitor for the window.")
    }
} else {
    print("No frontmost window found.")
}
```

You can compile it with(make sure to `chmod +x center_floats` when compiler is done):
```sh
swiftc center_floats.swift -o center_floats
```

Run it manually:
```sh
./center_floats -w 1600 -h 800
```

Or in aerospace.toml example:
```toml
f = [
  'exec-and-forget ~/.config/aerospace/scripts/center_floats',
  'move-mouse window-force-center',
  'mode main',
]
```

If you don't set a width(-w) and height(-h) it defaults to 1600x800, but you can change it before compiling it yourself if you would like. Also, this works as a toggle between floating and tiling layout. It just made sense to toggle back to tiling if you run it again and the window is already floating.

If you don't compile it you can still run the script by adding this line to the very top of the script or simply run with `swift center_floats.swift`:
```swift
#!/usr/bin/swift
```

Although I highly recommend compiling so it's not calling the interpreter each time.

ran a simple time comparison between the shell script I had vs the compiled swift script and got a noticeable difference every time:
```sh
time ./center_floats.sh
./center_floats.sh 0.11s user 0.30s system 58% cpu 0.693 total

time ./center_floats -w 1600 -h 900
./center_floats -w 1600 -h 900 0.01s user 0.01s system 14% cpu 0.154 total
```

**Reply by wojciech-kulik** [reply]
Why not just use some Rectangle, Magnet, Raycast or any other app that does that :D?

**Reply by gldtn** [reply]
You probably could.. but the goal is to have something integrated into Aerospace and not rely on other apps.

**Reply by wojciech-kulik** [reply]
That code is also another app :D But I get the point :)

**Comment by thimplicity** [comment]
@wojciech-kulik How would you "separate" rectangle and aerospace? I am new to aerospace and a central floating window is something I am missing.

**Reply by wojciech-kulik** [reply]
No need to separate, the apps that I listed are mostly moving windows around without tiling them, so you can just assign shortcuts to some operations like centering with a specific size.

**Reply by thimplicity** [reply]
Would aerospace not prevent that or would you define certain apps as floating in aerospace and then use another app to do the "centering"?

**Reply by wojciech-kulik** [reply]
Yea, the app must be floating to use another app for centering. You can achieve that however you like. I use floating by default because tiling is not very useful to me, I use most often one window per workspace

**Comment by monsterhxw** [comment]
I do this feature by using AppleScript to invoke Raycast window management center command hotkey.
Hope this can help others who find this later.

1. Modify config `~/.config/aerospace/aerospace.toml`
```toml
[mode.service.binding]
# Toggle between floating-centered and tiling layout.
## Floating-centered needs to depend on Raycast Window Management Center Command hotkey.
## My Raycast window management Center Command hotkey is set to `ctrl + option + cmd + f2`. (`Key Code 120` = F2)
period = ['''exec-and-forget aerospace layout floating && osascript -e 'tell application "System Events" to key code 120 using {control down, option down, command down}' || aerospace layout tiling''', 'mode main']
```
2. Set Hotkey
<img width="1112" alt="image" src="https://github.com/user-attachments/assets/fd0b3bac-3065-4f5a-b5bc-012b3489fe34" />

3. Done.

---

Or using shell script
```toml
[mode.service.binding]
# Toggle between floating-centered and tiling layout.
## Floating-centered needs to depend on Raycast Window Management Center Command hotkey.
## My Raycast window management Center Command hotkey is set to `ctrl + option + cmd + f2`. (`Key Code 120` = F2)
period = ['exec-and-forget ~/.config/aerospace/toggle-between-floating-centered-and-tiling-layout.sh', 'mode main']
```
`~/.config/aerospace/toggle-between-floating-centered-and-tiling-layout.sh`
```shell
#!/bin/bash 
aerospace layout floating \
  && osascript -e 'tell application "System Events" to key code 120 using {control down, option down, command down}' \
  || aerospace layout tiling
```

**Reply by zhuoqun-chen** [reply]
> I do this feature by using AppleScript to invoke Raycast window management center command hotkey.  Hope this can help others who find this later. 
> 
> 1. Modify config `~/.config/aerospace/aerospace.toml`修改配置
> 
> ```toml
> [mode.service.binding]
> # Toggle between floating-centered and tiling layout.
> ## Floating-centered needs to depend on Raycast Window Management Center Command hotkey.
> ## My Raycast window management Center Command hotkey is set to `ctrl + option + cmd + f2`. (`Key Code 120` = F2)
> period = ['''exec-and-forget aerospace layout floating && osascript -e 'tell application "System Events" to key code 120 using {control down, option down, command down}' || aerospace layout tiling''', 'mode main']
> ```
> 
> 2. Set Hotkey  设置热键
> 
> <img alt="image" width="1112" src="https://private-user-images.githubusercontent.com/38623561/402022078-fd0b3bac-3065-4f5a-b5bc-012b3489fe34.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDYxNjY0MTUsIm5iZiI6MTc0NjE2NjExNSwicGF0aCI6Ii8zODYyMzU2MS80MDIwMjIwNzgtZmQwYjNiYWMtMzA2NS00ZjVhLWI1YmMtMDEyYjM0ODlmZTM0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA1MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNTAyVDA2MDgzNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJiMDk4YzYyMDA1ODQyZDdkNzI1N2NiZTBjZmFmN2ZhN2Q0M2E5ZjllNTc4YWE0MjkzNmJjMjJlNTQzMWQ4ZTcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.NMmyasXqct2212G9oYkOTwFP26uDQ7uF_IdzD-Qg084">
> 3. Done.  完成。
> 
> Or using shell script或者使用 shell 脚本
> 
> ```toml
> [mode.service.binding]
> # Toggle between floating-centered and tiling layout.
> ## Floating-centered needs to depend on Raycast Window Management Center Command hotkey.
> ## My Raycast window management Center Command hotkey is set to `ctrl + option + cmd + f2`. (`Key Code 120` = F2)
> period = ['exec-and-forget ~/.config/aerospace/toggle-between-floating-centered-and-tiling-layout.sh', 'mode main']
> ```
> 
> `~/.config/aerospace/toggle-between-floating-centered-and-tiling-layout.sh`
> 
> ```shell
> #!/bin/bash 
> aerospace layout floating \
>   && osascript -e 'tell application "System Events" to key code 120 using {control down, option down, command down}' \
>   || aerospace layout tiling
> ```

thx, this works for me!

**Comment by LintaoAmons** [comment]
I could also do this with hammerspoon scripts, but don't want to have one extra tool to achieve it.
What's the best way to do it with aerospace?

**Comment by itpropro** [comment]
This is quite an essential functionality, what is the native way to achieve a centered floating window @nikitabobko ?

**Reply by nikitabobko** [reply]
This won't move forward until somebody submits a reasonable synopsis design proposal for the subcommand that could manipulate floating windows to the #feature-ideas. The proposal has to cover a wide variety of manipulation with floating windows, not only centering.

I personally don't care about floating windows as use them only for temporary dialogs. AeroSpace is a *tiling* window manager after all. That's why I am not spending my brain power on this issue.

**Comment by diegodorado** [comment]
I got a workaround that works pretty well with this:

```
# open example, set title to "scratch"
open -na WezTerm --args start --always-new-process -- bash -lc 'printf "\033]0;scratch\007"; fzf | pbcopy'
```

and then on aerospace config:

```
[[on-window-detected]]
if.app-id = "com.github.wez.wezterm"
if.window-title-regex-substring = 'scratch'
run = "layout floating"
```

**Comment by alad** [comment]
If you use raycast (pro), here's a solution that combines the best of aerospace and raycast:

```
    # Float and center the window. If it's already floating, return it back into the tiles.
    alt-i = """exec-and-forget aerospace layout floating && sleep 0.2 && open -g 'raycast://customWindowManagementCommand?&name=Center%20Focus&position=center&relativeWidth=0.6&relativeHeight=0.7' || aerospace layout tiling"""
```

**Reply by fgeck** [reply]
For me your command requires Raycast Pro.
Maybe add this as a note to your proposal

**Comment by mufeedcm** [comment]
for those who just want it to open floating, like i need emacs to show the todos and all, 
just add, 

```
[[on-window-detected]]
if.app-id = "org.gnu.Emacs"
run = "layout floating"

```
this to the end of your aerospace.toml

you can replace the app id with the app id of the app you need, you can get it by running ` osascript -e 'id of app "Application-Name"' `

this won't be able to set the sizing or center it, 
but it will open it almost at the center and the size is enough for me,

<img width="1440" height="900" alt="image" src="https://github.com/user-attachments/assets/3fbc0d57-8cc1-4eba-8641-8cc53134577b" />

**Comment by vieitesss** [comment]
What I ended up doing was:
- A key bind to toggle floating mode.
- Use hammerspoon to place the window at the center when floating.

Here https://github.com/vieitesss/.dot/tree/main/config you have my configuration.
